### PR TITLE
Add range, slice function in vector module

### PR DESCRIPTION
**Pull request by JohnChangUK**
### Description
Adding the `range` and `slice` functions inside the `vector` module. 
### `range`
- The `range` function generates a sequence of `u64` integers from a specified start to an end (exclusive), incremented by a given step.
- These are unsigned integers, therefore we don't need to handle negative number cases.

### `slice`
- The `slice` function takes a reference vector, specified start to an end (exclusive) and returns a new vector with the copied values from `start` to `end` index.

### Test Plan
To run tests
1.  `cd` into directory `aptos-core/aptos-move/framework/move-stdlib`
2. run `aptos move test --filter test_range`
3. run `aptos move test --filter test_slice`

### Range test cases
1. Happy case
4. Edge case where `start` and `end` are equal - (Should return empty vector)
5. Edge case where `end` is greater than `start`  - (Should return empty vector)
6. If `start` is 0, abort as this results in an infinite loop, wasting all gas provided in the transaction.

### Slice test cases
1. Happy case - start and end portion of vector
2. Empty vector cases
3. Full slice case.
4. Invalid `start` test case
5. Invalid `end` test case

### CODE REVIEW

File: aptos-move/framework/move-stdlib/sources/vector.move [hunk 1]
```diff
@@ -589,6 +592,19 @@ module std::vector {
         for_each_reverse(v, |e| d(e))
     }
 
+    public inline fun range(start: u64, end: u64, step: u64): vector<u64> {
```
Comment (movekevin) [inline aptos-move/framework/move-stdlib/sources/vector.move:595]: I wonder if it makes sense to have 2 separate range functions - one with step and one without step (default to 1). Also IMO this should not be an inline function as it's long enough and would just bloat people's bytecodes if used multiple times in the same module - this entire code would get expanded a few times.

Comment (JohnChangUK) [reply aptos-move/framework/move-stdlib/sources/vector.move:595]: Good point, will remove the `inline` here. I'll create another function named `range_with_single_step`.

File: aptos-move/framework/move-stdlib/sources/vector.move [hunk 2]
```diff
@@ -589,6 +592,21 @@ module std::vector {
         for_each_reverse(v, |e| d(e))
     }
 
+    public fun range(start: u64, end: u64): vector<u64> {
```
Comment (movekevin) [inline aptos-move/framework/move-stdlib/sources/vector.move:595]: You need a mut reference to the vector for both the functions

File: aptos-move/framework/move-stdlib/sources/vector.move [hunk 3]
```diff
@@ -589,6 +592,21 @@ module std::vector {
...
+        range_with_step(start, end, 1)
+    }
+
+    public fun range_with_step(start: u64, end: u64, step: u64): vector<u64> {
+        assert!(step > 0, EINVALID_STEP);
+
+        let vec = vector<u64>[];
```
[the remark this answers is not in the document]
Comment (JohnChangUK) [reply aptos-move/framework/move-stdlib/sources/vector.move:602]: Ack, resolved.

File: aptos-move/framework/move-stdlib/sources/vector.move (hunk 2 above)
Comment (JohnChangUK) [reply aptos-move/framework/move-stdlib/sources/vector.move:595]: Sounds good, will add.

**Review by movekevin — APPROVED** [review]

**Review by junkil-park — COMMENTED** [review]
This looks good! Can we also think about whether we would also need `range` for different number types (u8, u256)?  Potentially having `fun range<T>(...): vector<T>`?

**Comment by JohnChangUK** [comment]
> This looks good! Can we also think about whether we would also need `range` for different number types (u8, u256, ...)?

As we discussed, the type parameter wouldn't work because we can't constrain T to be numeric types.

**Review by alinush — APPROVED** [review]

**Pull request merged**